subsys/bluetooth/host/l2cap.c

```diff
@@ -930,13 +929,20 @@ static void l2cap_chan_tx_process(struct k_work *work)
 
 	/* Resume tx in case there are buffers in the queue */
 	while ((buf = l2cap_chan_le_get_tx_buf(ch))) {
+		uint16_t rem_len = buf->len;
+
 		/* Here buf is either:
 		 * - a partially-sent SDU le_chan->tx_buf
 		 * - a new SDU from the TX queue
 		 */
 		LOG_DBG("chan %p buf %p", ch, buf);
 
-		ret = l2cap_chan_le_send_sdu(ch, buf);
+		ret = l2cap_chan_le_send(ch, buf);
+
+		if (ret == rem_len) {
+			net_buf_unref(buf);
+		}
```

Comment:
There is some insane ownership logic going on here. `l2cap_chan_le_send` takes ownership of `buf` unless error or `ret == rem_len`. @jori-nordic Can you confirm I did this refactor right? I'm struggling to follow the `tx_buf` logic.

Reply:
The loop does loop :/
I will refactor this when I have the time (lol)